test/unit/space_index_test.lua

```diff
@@ -0,0 +1,301 @@
+
+    tmp_space = box.schema.create_space(space_name)
+    tmp_space:create_index("primary", {type = "TREE", parts = {{field = 1}}})
+    tmp_space:insert({2, "2"})
+
+    helpers.retrying({}, function()
+        t.assert_equals(helpers.iteration_result, {{2, "2"}})
+    end)
+    t.assert_equals(#expirationd.tasks(), 1)
+    t.assert_not_equals(expirationd.task(task_name), nil)
+    t.assert_equals(task:statistics().restarts, 1)
+
+    task:kill()
+end
+
+function g.test_run_after_non_existing_space_and_index_created()
+    local task_name = "clean_all"
+    local space_name = "tmp"
+    local index_name = "non_primary"
+
+    helpers.iteration_result = {}
+    local task = expirationd.start(task_name, space_name, helpers.is_expired_debug,
+                                   {index = index_name})
+
+    tmp_space = box.schema.create_space(space_name)
+    tmp_space:create_index("primary", {type = "TREE", parts = {{field = 1}}})
+    tmp_space:insert({2, "2"})
+
+    fiber.sleep(double_worker_restart_time)
+
+    t.assert_equals(helpers.iteration_result, {})
+    t.assert_equals(#expirationd.tasks(), 1)
+    t.assert_not_equals(expirationd.task(task_name), nil)
+    t.assert_equals(task:statistics().restarts, 0)
+
+    tmp_space:create_index(index_name, {type = "TREE", parts = {{field = 1}}})
+
+    helpers.retrying({}, function()
+        t.assert_equals(helpers.iteration_result, {{2, "2"}})
+    end)
+    t.assert_equals(#expirationd.tasks(), 1)
+    t.assert_not_equals(expirationd.task(task_name), nil)
+    t.assert_equals(task:statistics().restarts, 1)
+
+    task:kill()
+end
+
+local rename_cases = {
+    index_rename = {
+        fun = function(space, _)
+            space:rename("XXX")
+        end
+    },
+    space_rename = {
+        fun = function(_, index)
+            index:rename("XXX")
+        end
+    },
+}
+
+for name, case in pairs(rename_cases) do
+    g["test_stop_after_" .. name .. "_if_name"] = function(cg)
+        local task_name = "clean_all"
+        local space_name = cg.space.name
+        local index_name = "index_for_first_name"
+
+        helpers.iteration_result = {}
+        local task = expirationd.start(task_name, space_name, helpers.is_expired_debug,
+                                       {index = index_name})
+        helpers.retrying({}, function()
+            t.assert_equals(helpers.iteration_result, {{1, "1"}})
+        end)
+        helpers.iteration_result = {}
+
+        local capture = luatest_capture:new()
+        capture:wrap(true, function()
+            local space = box.space[space_name]
+            case.fun(space, space.index[index_name])
+            space:insert({1, "1"})
+
+            local stderr = ""
+            helpers.retrying({}, function()
+                stderr = stderr .. capture:flush().stderr
+                t.assert_str_contains(stderr, "expiration: stop task")
+            end)
+        end)
+
+        t.assert_equals(helpers.iteration_result, {})
```

Comment:
How should one interpret this block of asserts?

Reply:
The task doesn't process any tuple.

The pattern `helpers.iteration_result + helpers.is_expired_debug` used in many tests.

Reply:
The task has run a single time in
```lua
helpers.retrying({}, function()
    t.assert_equals(helpers.iteration_result, {{1, "1"}})
end)
```

How we can differ between "the task has stopped" and "timeout is not yet passed for a task to run a second time, but it will run if we wait a little bit"?

Reply:
There is no way to check the state of the task from an user side. It remains only to believe the message in the logs.

Reply:
> There is no way to check the state of the task from an user side.

Well, that's a pity. But I still don't get why in the end of both `"test_stop_after_" .. name .. "_if_name_used"` and `"test_not_stop_after_" .. name .. "if_id_used"` we assert the same `t.assert_equals(task:statistics().restarts, 1)`

Reply:
I've added:
```
t.assert_equals(task.worker_fiber:status(), "dead")
t.assert_equals(task.guardian_fiber:status(), "dead")
```
into `*_stop_*` tests. These are implementation details, but we should quickly notice if changes in the implementation break our tests (we will rename the fibers, or delete a fiber).

> Well, that's a pity. But I still don't get why in the end of both "test_stop_after_smth_if_name_used" and "test_not_stop_after_smth_if_id_used" we assert the same `t.assert_equals(task:statistics().restarts, 1)`

restarts == 0 if a task execution started, restarts == 1 if a task execution started, restarts == 2 if a task execution started again (an error happens).

Reply:
>These are implementation details

I think it's okay for a unit test.